core/runtime_modules.rs

```diff
@@ -0,0 +1,206 @@
+  js_files: Option<Vec<SourcePair>>,
+  ops: Option<Vec<OpPair>>,
+  opstate_fn: Option<Box<OpStateFn>>,
+}
+
+impl BasicModule {
+  pub fn new(
+    js_files: Option<Vec<SourcePair>>,
+    ops: Option<Vec<OpPair>>,
+    opstate_fn: Option<Box<OpStateFn>>,
+  ) -> Self {
+    Self {
+      js_files,
+      ops,
+      opstate_fn,
+    }
+  }
+
+  pub fn pure_js(js_files: Vec<SourcePair>) -> Self {
+    Self::new(Some(js_files), None, None)
+  }
+
+  pub fn with_ops(
+    js_files: Vec<SourcePair>,
+    ops: Vec<OpPair>,
+    opstate_fn: Option<Box<OpStateFn>>,
+  ) -> Self {
+    Self::new(Some(js_files), Some(ops), opstate_fn)
+  }
+}
+
+impl JsRuntimeModule for BasicModule {
+  fn init_js(&self) -> Result<Vec<SourcePair>, AnyError> {
+    Ok(match &self.js_files {
+      Some(files) => files.clone(),
+      None => vec![],
+    })
+  }
+
+  fn init_ops(&mut self, registrar: RcOpRegistrar) -> Result<(), AnyError> {
+    // NOTE: not idempotent
+    // TODO: fail if called twice ?
+    if let Some(ops) = self.ops.take() {
+      for (name, opfn) in ops {
+        registrar.borrow_mut().register_op(name, opfn);
+      }
+    }
+    Ok(())
+  }
+
+  fn init_state(&self, state: &mut OpState) -> Result<(), AnyError> {
+    match &self.opstate_fn {
+      Some(ofn) => ofn(state),
+      None => Ok(()),
+    }
+  }
+}
+
+// MultiModule allows grouping multiple sub-JsRuntimeModules into one,
+// allowing things such as:
+// ```
+// fn web_modules(args: WebModuleArgs) -> MultiModule {
+//  MultiModule::new(vec![deno_url::init(), deno_fetch::init(...), ...])
+// }
+// ```
+pub struct MultiModule<'s> {
+  pub modules: Vec<Box<dyn JsRuntimeModule + 's>>,
+}
+
+impl MultiModule<'_> {
+  fn new<'s>(modules: &mut Vec<impl JsRuntimeModule + 's>) -> MultiModule<'s> {
+    let modules = modules
+      .drain(..)
+      .map(|m| Box::<dyn JsRuntimeModule + 's>::from(Box::new(m)))
+      .collect();
+    MultiModule { modules }
+  }
+}
+
+impl JsRuntimeModule for MultiModule<'_> {
+  fn init_js(&self) -> Result<Vec<SourcePair>, AnyError> {
+    Ok(
+      self
+        .modules
+        .iter()
+        .map(|m| m.init_js().unwrap())
+        .flatten()
+        .collect(),
+    )
+  }
+
+  fn init_ops(&mut self, registrar: RcOpRegistrar) -> Result<(), AnyError> {
+    for m in self.modules.iter_mut() {
+      m.init_ops(registrar.clone())?;
+    }
+    Ok(())
+  }
+}
+
+// The OpRegistrar trait allows building op "middleware" such as:
+// OpMetrics, OpTracing or OpDisabler that wrap OpFns for profiling, debugging, etc...
+// JsRuntime is itself an OpRegistrar
+pub trait OpRegistrar {
+  fn register_op(&mut self, name: &str, op_fn: Box<OpFn>) -> OpId;
+  // register_minimal_op_sync(...)
+  // register_minimal_op_async(...)
+  // register_json_op_sync(...)
+  // register_json_op_async(...)
+}
+
+////
+// Helper macros to reduce verbosity / redundant decls
+////
+#[macro_export]
+macro_rules! include_js_files {
+  (root $root:literal, $($file:literal,)+) => {
```

Comment:
How about renaming `root` to `prefix`? My first thought was that it tell a root path to resolve files, while actually it's the prefix that will be added to declared paths when executing files.

Reply:
An example in the docstring would be helpful

Reply:
@bartlomieju Are these good enough ? https://github.com/denoland/deno/blob/dcdd86c74eea328d6fdad78d5ffef7e65fb6d004/core/runtime_modules.rs#L242-L303

(also renamed `root` => `prefix`)